### PR TITLE
[Navigation] Fix cross-document navigations in subframes

### PR DESCRIPTION
**Pull request by rwlbuis**
#### 61ed07908d96469ff94ac923699ce5cd53cb58a0
<pre>
[Navigation] Fix cross-document navigations in subframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=280750">https://bugs.webkit.org/show_bug.cgi?id=280750</a>

Reviewed by Alex Christensen.

Fix cross-document navigations done on subframes, we need extra logic to not lose the NavigationHistoryEntry bookkeeping done in the previous window.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-traverseTo-navigates-top-and-same-doc-child-and-cross-doc-child-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-traverseTo-top-cancels-cross-document-child-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-push-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-same-document-then-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-blank-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-cross-document-forward-pruning-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-javascript-url-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-navigations-in-multiple-windows-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-srcdoc-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/no-referrer-dynamic-url-censored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/no-referrer-from-meta-url-censored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/no-referrer-url-censored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-history-push-same-url-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-already-detached-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-detach-cross-document-before-navigate-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/sandboxing-back-parent-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/sandboxing-back-sibling-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/cross-document-getState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/cross-document-getState-undefined-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/cross-document-location-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/cross-document-location-api-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/page/Navigation.cpp:
(WebCore::getEntryIndexOfHistoryItem):
(WebCore::Navigation::initializeForNewWindow):
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::create):

Canonical link: <a href="https://commits.webkit.org/284785@main">https://commits.webkit.org/284785@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/a21dc0d3ebfbe3ee97553a888f9c3142a564eb27

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70504 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49910 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23269 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74593 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21682 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72621 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57710 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21522 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55865 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/14336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73570 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45393 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60779 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36327 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42052 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18215 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20043 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63984 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18560 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76312 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14730 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17794 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63589 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14773 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60846 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63523 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15611 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11564 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5205 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45712 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46786 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48063 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46528 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/page/Navigation.cpp [hunk 1]
```diff
@@ -122,14 +122,16 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
     RefPtr currentItem = frame()->history().currentItem();
     if (!currentItem)
         return;
-    if (previousWindow) {
+    if (previousWindow && !frame()->isMainFrame()) {
```
Comment (TingPing) [inline Source/WebCore/page/Navigation.cpp:124]: Could you elaborate on what is special about the mainFrame?

File: Source/WebCore/page/Navigation.cpp [hunk 2]
```diff
@@ -93,12 +93,12 @@ bool Navigation::canGoForward() const
...
 
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#getting-the-navigation-api-entry-index
-static std::optional<size_t> getEntryIndexOfHistoryItem(const Vector<Ref<NavigationHistoryEntry>>& entries, const HistoryItem& item)
+static std::optional<size_t> getEntryIndexOfHistoryItem(const Vector<Ref<NavigationHistoryEntry>>& entries, const HistoryItem& item, size_t start = 0)
 {
     size_t index = 0;
     // FIXME: We could have a more efficient solution than iterating through a list.
     for (auto& entry : entries) {
-        if (entry->associatedHistoryItem() == item)
+        if (entry->associatedHistoryItem() == item && index >= start)
```
Comment (TingPing) [inline Source/WebCore/page/Navigation.cpp]: This is starting to diverge from the spec, which I understand we need to regarding cross-document histories, but I'd like to see more comments documenting the differences.

File: Source/WebCore/page/Navigation.cpp (hunk 1 above)
Comment (rwlbuis) [reply Source/WebCore/page/Navigation.cpp:124]: I added a comment.

File: Source/WebCore/page/Navigation.cpp (hunk 2 above)
Comment (rwlbuis) [reply Source/WebCore/page/Navigation.cpp]: Yes, it is unfortunate. I considered doing a reverse traversal, but I don't think that is much better if at all.

File: Source/WebCore/page/Navigation.cpp [hunk 3]
```diff
@@ -122,14 +122,17 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
     RefPtr currentItem = frame()->history().currentItem();
     if (!currentItem)
         return;
-    if (previousWindow) {
+    // For main frames we can still rely on the page b/f list. However for subframes we need below logic to not lose the bookkeeping done in the previous window.
+    if (previousWindow && !frame()->isMainFrame()) {
         Ref previousNavigation = previousWindow->protectedNavigation();
-        if (previousNavigation->m_entries.size() && navigationType && navigationType == NavigationNavigationType::Reload && frame()->document()->protectedSecurityOrigin()->isSameOriginDomain(previousWindow->document()->protectedSecurityOrigin())) {
+        if (previousNavigation->m_entries.size() && navigationType && (navigationType == NavigationNavigationType::Reload || navigationType == NavigationNavigationType::Push) && frame()->document()->protectedSecurityOrigin()->isSameOriginDomain(previousWindow->document()->protectedSecurityOrigin())) {
```
Comment (TingPing) [inline Source/WebCore/page/Navigation.cpp]:
```suggestion
        if (previousNavigation->m_entries.size() && (navigationType == NavigationNavigationType::Reload || navigationType == NavigationNavigationType::Push) && frame()->document()->protectedSecurityOrigin()->isSameOriginDomain(previousWindow->document()->protectedSecurityOrigin())) {
```

Comment (TingPing) [reply Source/WebCore/page/Navigation.cpp]: Are you sure about SameOriginDomain instead of just SameOrigin?

Comment (rwlbuis) [reply Source/WebCore/page/Navigation.cpp]: Fixed.

**Comment by rwlbuis** [comment]
Note that most of the remaining FAILs in this patch are due to not handling subframe history traversals correctly, that is probably the next thing we need to address.

File: Source/WebCore/page/Navigation.cpp (hunk 2 above)
Comment (achristensen07) [inline Source/WebCore/page/Navigation.cpp]: Instead of adding this constraint and unnecessarily iterating entries before the start just to do a comparison and increment an integer, let's start at start and only do a comparison on the possibilities.  No c++11-style iteration in this case.

Alternatively, we could pass in a std::span of Refs and have the caller do a subspan instead of passing in a vector and start.

File: Source/WebCore/page/Navigation.cpp [hunk 4]
```diff
@@ -122,14 +122,17 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
     RefPtr currentItem = frame()->history().currentItem();
     if (!currentItem)
         return;
-    if (previousWindow) {
+    // For main frames we can still rely on the page b/f list. However for subframes we need below logic to not lose the bookkeeping done in the previous window.
+    if (previousWindow && !frame()->isMainFrame()) {
         Ref previousNavigation = previousWindow->protectedNavigation();
-        if (previousNavigation->m_entries.size() && navigationType && navigationType == NavigationNavigationType::Reload && frame()->document()->protectedSecurityOrigin()->isSameOriginDomain(previousWindow->document()->protectedSecurityOrigin())) {
+        if (previousNavigation->m_entries.size() && (navigationType == NavigationNavigationType::Reload || navigationType == NavigationNavigationType::Push) && frame()->document()->protectedSecurityOrigin()->isSameOriginAs(previousWindow->document()->protectedSecurityOrigin())) {
```
Comment (achristensen07) [inline Source/WebCore/page/Navigation.cpp]: This looks a bit messy and hard to follow, especially since there is a check for NavigationNavigationType::Reload inside.  Instead, please use a switch statement or make a function that has a name that indicates what it is about ::Reload and ::Push that meets the condition here, or otherwise make the logic easier to follow.

Also, is the switch from isSameOriginDomain to isSameOriginAs intentional?

**Review by achristensen07 — CHANGES_REQUESTED** [review]

File: Source/WebCore/page/Navigation.cpp (hunk 4 above)
Comment (TingPing) [reply Source/WebCore/page/Navigation.cpp]: I believe SameOrigin is correct:

> newSHEs will have originally come from [getting session history entries for the navigation API](https://html.spec.whatwg.org/multipage/browsing-the-web.html#getting-session-history-entries-for-the-navigation-api), and thus each newSHE will be contiguous [same](https://html.spec.whatwg.org/multipage/browsers.html#same-origin) [origin](https://html.spec.whatwg.org/multipage/browsing-the-web.html#document-state-origin) with initialSHE.

File: Source/WebCore/page/Navigation.cpp (hunk 2 above)
Comment (rwlbuis) [reply Source/WebCore/page/Navigation.cpp]: I rewrote to use a subspan internally. When trying to hand a suitable span as parameter the index counting did not take the skipped entries into account so the calculated m_currentEntryIndex was off.

File: Source/WebCore/page/Navigation.cpp (hunk 4 above)
Comment (rwlbuis) [reply Source/WebCore/page/Navigation.cpp]: I added a lambda for readability.

File: Source/WebCore/page/Navigation.cpp [hunk 5]
```diff
@@ -93,13 +93,13 @@ bool Navigation::canGoForward() const
 }
 
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#getting-the-navigation-api-entry-index
-static std::optional<size_t> getEntryIndexOfHistoryItem(const Vector<Ref<NavigationHistoryEntry>>& entries, const HistoryItem& item)
+static std::optional<size_t> getEntryIndexOfHistoryItem(const Vector<Ref<NavigationHistoryEntry>>& entries, const HistoryItem& item, size_t start = 0)
 {
     size_t index = 0;
     // FIXME: We could have a more efficient solution than iterating through a list.
-    for (auto& entry : entries) {
+    for (auto& entry : entries.subspan(start)) {
```
Comment (achristensen07) [inline Source/WebCore/page/Navigation.cpp]: I think this would be neater like this:
for (size_t index = start; index < entries.size(); index++)

File: Source/WebCore/page/Navigation.cpp [hunk 6]
```diff
@@ -122,14 +122,26 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
...
         return;
-    if (previousWindow) {
+    // For main frames we can still rely on the page b/f list. However for subframes we need below logic to not lose the bookkeeping done in the previous window.
+    if (previousWindow && !frame()->isMainFrame()) {
         Ref previousNavigation = previousWindow->protectedNavigation();
-        if (previousNavigation->m_entries.size() && navigationType && navigationType == NavigationNavigationType::Reload && frame()->document()->protectedSecurityOrigin()->isSameOriginDomain(previousWindow->document()->protectedSecurityOrigin())) {
+        bool shouldProcessPreviousNavigationEntries = [&]() {
+            if (!previousNavigation->m_entries.size())
+                return false;
+            if (!(navigationType == NavigationNavigationType::Reload || navigationType == NavigationNavigationType::Push))
```
Comment (achristensen07) [inline Source/WebCore/page/Navigation.cpp]: if (a != b && a != c) rather than if (!(a == b || a == c)) would be easier to read, I think.

**Review by achristensen07 — APPROVED** [review]

File: Source/WebCore/page/Navigation.cpp (hunk 6 above)
Comment (rwlbuis) [reply Source/WebCore/page/Navigation.cpp]: Done.

File: Source/WebCore/page/Navigation.cpp (hunk 5 above)
Comment (rwlbuis) [reply Source/WebCore/page/Navigation.cpp]: Done.
